### PR TITLE
Initial implementation of the final step towards cross-version support.

### PR DESCRIPTION
**Pull request by rhc54**
This current state only supports message-based data sharing - it specifically does not support dstore at this time as the relevant components need to be implemented.

It adds three new frameworks:

* bfrops - handles the value packing/unpacking functions as there are slight variations between the major PMIx versions. Only the v2.x component has been tested at this time.

* gds - generalized data store. The components here provide support for the hash table/message method of sharing client/server data, plus the shared memory methods (with different components for v1.1 and v1.2 since the locking support changed). Only the hash component has been implemented here.

* pshmem - various shared memory support frameworks, equivalent to the "esh" support in the current dstore implementation.

Note that macros are used to access the GDS, BFROPS, and PTL functions to simplify the interfaces.

We can track the required additional steps here:

- [x] Update the usock ptl component
- [ ] Update all frameworks to use their own verbosity instead of globals.debug
- [x] Implement dstore v2.x support
- [ ] Implement dstore v1.2 support
- [ ] Implement dstore v1.1 support
- [ ] Implement bfrops v1.x support
- [ ] Install in PMIx reference server
- [ ] Test against OMPI v2.x clients
- [ ] Test against OMPI master clients
- [ ] Check fence, get, connect, disconnect across versions
- [ ] Check tool support
- [ ] Check event notification across versions
- [ ] Install in OMPI and test

Signed-off-by: Ralph Castain <rhc@open-mpi.org>

### CODE REVIEW

**Comment by rhc54** [comment]
@artpol84 I don't understand what test is failing - it saids something about:

 ```
16:18:15 + test_id=15
16:18:15 + '[' no = yes ']'
16:18:15 + '[' 1 = 0 ']'
16:18:15 + echo 'Test failed'
16:18:15 Test failed
```
Can you explain?

**Comment by artpol84** [comment]
@rhc54 you need to scroll up:

```
02:15:01 + eval timeout -s SIGKILL 1m ./pmix_test -n 5 --test-resolve-peers --ns-dist '"1:2:2"' -o '$OUTDIR/out'
02:15:01 ++ timeout -s SIGKILL 1m ./pmix_test -n 5 --test-resolve-peers --ns-dist 1:2:2 -o /var/lib/jenkins/jobs/gh-pmix-pr/workspace/out/out
02:15:01 [jenkins03:05773] PMIX ERROR: OUT-OF-RESOURCE in file /var/lib/jenkins/jobs/gh-pmix-pr/workspace/src/server/pmix_server_ops.c at line 992
02:15:01 ERROR [cli_stages.c:256:errhandler]:  PMIX server event handler with status = -25
02:15:01 ERROR [cli_stages.c:148:cli_cleanup]: rank 0 with state 1 unexpectedly terminated.
02:15:01 ERROR [test_resolve_peers.c:38:resolve_nspace]: smoky_nspace-1:2: Resolve peers returned incorect result: returned 1 processes, expected 2
02:15:01 ERROR [test_resolve_peers.c:67:test_resolve_peers]: smoky_nspace-1:2: Resolve peers failed for the own namespace
02:15:01 
02:15:01 ERROR [pmix_client.c:183:main]: smoky_nspace-1:2 Resolve peers test failed: -1
02:15:01 ERROR [pmix_test.c:176:main]: Test was aborted!
```

```
02:15:01 + eval timeout -s SIGKILL 1m ./pmix_test -n 2 --test-connect -o '$OUTDIR/out'
02:15:01 ++ timeout -s SIGKILL 1m ./pmix_test -n 2 --test-connect -o /var/lib/jenkins/jobs/gh-pmix-pr/workspace/out/out
02:15:01 ERROR [test_cd.c:71:test_connect_disconnect]: smoky_nspace-0:0: Connect blocking test failed.
02:15:01 ERROR [pmix_client.c:174:main]: smoky_nspace-0:0 Connect/Disconnect test failed: -1
02:15:01 ERROR [cli_stages.c:148:cli_cleanup]:  PMIX0 with state 1 unexpectedly terminated.
02:15:01 ERROR [cli_stages.c:256:errhandler]:  PMIX0 with state 1 unexpectedly terminated.
02:15:01 ERROR [test_cd.c:71:test_connect_disconnect]: smoky_nspace-0:1: Connect blocking test failed.
02:15:01 ERROR [pmix_client.c:174:main]: smoky_nspace-0:1 Connect/Disconnect test failed: -1
02:15:01 ERROR [pmix_test.c:176:main]: Test was aborted!
```

```
02:15:01 ++ timeout -s SIGKILL 1m ./pmix_test -n 2 --test-spawn -o /var/lib/jenkins/jobs/gh-pmix-pr/workspace/out/out
02:15:01 [jenkins03:05685] PMIX ERROR: UNPACK-PAST-END in file /var/lib/jenkins/jobs/gh-pmix-pr/workspace/src/client/pmix_client_spawn.c at line 216
02:15:01 [jenkins03:05686] PMIX ERROR: UNPACK-PAST-END in file /var/lib/jenkins/jobs/gh-pmix-pr/workspace/src/client/pmix_client_spawn.c at line 216
02:15:01 [jenkins03:05685] PMIX ERROR: UNPACK-PAST-END in file /var/lib/jenkins/jobs/gh-pmix-pr/workspace/src/client/pmix_client.c at line 114
02:15:01 [jenkins03:05686] PMIX ERROR: UNPACK-PAST-END in file /var/lib/jenkins/jobs/gh-pmix-pr/workspace/src/client/pmix_client.c at line 114
02:15:01 ERROR [pmix_client.c:52:errhandler]: PMIX client: Error handler with status = -101
02:15:01 ERROR [pmix_client.c:52:errhandler]: PMIX client: Error handler with status = -101
02:15:01 ERROR [test_spawn.c:81:test_spawn]: smoky_nspace-0:0: Spawn non-blocking test failed.
02:15:01 Segmentation fault
02:15:01 ERROR [pmix_client.c:165:main]: smoky_nspace-0:0 Spawn test failed: -1
02:15:01 ERROR [test_spawn.c:81:test_spawn]: smoky_nspace-0:1: Spawn non-blocking test failed.
02:15:01 + ret=139
02:15:01 ERROR [pmix_client.c:165:main]: smoky_nspace-0:1 Spawn test failed: -1
```

```
02:15:00 + eval timeout -s SIGKILL 1m ./pmix_test -n 2 --test-publish -o '$OUTDIR/out'
02:15:00 ++ timeout -s SIGKILL 1m ./pmix_test -n 2 --test-publish -o /var/lib/jenkins/jobs/gh-pmix-pr/workspace/out/out
02:15:01 [jenkins03:05636] PMIX ERROR: BAD-PARAM in file /var/lib/jenkins/jobs/gh-pmix-pr/workspace/src/mca/bfrops/base/bfrop_base_pack.c at line 46
02:15:01 [jenkins03:05636] PMIX ERROR: BAD-PARAM in file /var/lib/jenkins/jobs/gh-pmix-pr/workspace/src/client/pmix_client_pub.c at line 307
02:15:01 ERROR [test_publish.c:159:test_publish_lookup_common]: smoky_nspace-0:1: PMIX_Lookup failed.
02:15:01 ERROR [test_publish.c:185:test_publish_lookup]: smoky_nspace-0:1: Publish/Lookup blocking test failed.
02:15:01 ERROR [pmix_client.c:156:main]: smoky_nspace-0:1 Publish/Lookup test failed: -1
02:15:01 ERROR [cli_stages.c:256:errhandler]:  PMIX server event handler with status = -25
02:15:01 ERROR [cli_stages.c:148:cli_cleanup]: rank 1 with state 1 unexpectedly terminated.
02:15:01 ERROR [pmix_test.c:176:main]: Test was aborted!
```
```
02:15:00 + check_result 'blocking fence w/o data on the same nspace' './pmix_test -n 4 --job-fence -o $OUTDIR/out'
02:15:00 + set +e
02:15:00 + eval timeout -s SIGKILL 1m ./pmix_test -n 4 --job-fence -o '$OUTDIR/out'
02:15:00 ++ timeout -s SIGKILL 1m ./pmix_test -n 4 --job-fence -o /var/lib/jenkins/jobs/gh-pmix-pr/workspace/out/out
02:15:00 [jenkins03:05536] PMIX ERROR: UNPACK-FAILURE in file /var/lib/jenkins/jobs/gh-pmix-pr/workspace/src/server/pmix_server_ops.c at line 197
02:15:00 [jenkins03:05536] PMIX ERROR: UNPACK-FAILURE in file /var/lib/jenkins/jobs/gh-pmix-pr/workspace/src/server/pmix_server_ops.c at line 197
02:15:00 [jenkins03:05536] PMIX ERROR: UNPACK-FAILURE in file /var/lib/jenkins/jobs/gh-pmix-pr/workspace/src/server/pmix_server_ops.c at line 197
02:15:00 ERROR [test_fence.c:385:test_job_fence]: smoky_nspace-0:0: PMIx_Get failed: -1
02:15:00 ERROR [pmix_client.c:147:main]: smoky_nspace-0:0 Job fence test failed: -1
02:15:00 [jenkins03:05536] PMIX ERROR: UNPACK-FAILURE in file /var/lib/jenkins/jobs/gh-pmix-pr/workspace/src/server/pmix_server_ops.c at line 197
02:15:00 ERROR [cli_stages.c:256:errhandler]:  PMIX server event handler with status = -25
02:15:00 ERROR [cli_stages.c:148:cli_cleanup]: rank 0 with state 1 unexpectedly terminated.
02:15:00 ERROR [pmix_test.c:176:main]: Test was aborted!
02:15:00 Segmentation fault
```

```
02:14:50 ++ timeout -s SIGKILL 1m ./pmix_test -n 4 --job-fence -c -o /var/lib/jenkins/jobs/gh-pmix-pr/workspace/out/out
02:14:50 [jenkins03:05450] PMIX ERROR: UNPACK-FAILURE in file /var/lib/jenkins/jobs/gh-pmix-pr/workspace/src/server/pmix_server_ops.c at line 197
02:14:50 [jenkins03:05450] PMIX ERROR: UNPACK-FAILURE in file /var/lib/jenkins/jobs/gh-pmix-pr/workspace/src/server/pmix_server_ops.c at line 197
02:14:50 [jenkins03:05450] PMIX ERROR: UNPACK-FAILURE in file /var/lib/jenkins/jobs/gh-pmix-pr/workspace/src/server/pmix_server_ops.c at line 197
02:14:50 [jenkins03:05450] PMIX ERROR: UNPACK-FAILURE in file /var/lib/jenkins/jobs/gh-pmix-pr/workspace/src/server/pmix_server_ops.c at line 197
02:15:00 ERROR [pmix_test.c:170:main]: Test exited by a timeout!
02:15:00 ERROR [cli_stages.c:148:cli_cleanup]: rank 0 with state 1 unexpectedly terminated.
02:15:00 ERROR [pmix_test.c:176:main]: Test was aborted!
```

```
02:14:40 ++ timeout -s SIGKILL 1m ./pmix_test -n 4 --ns-dist 3:1 --fence '[d | 0:]' --noise '[0:0,1]' -o /var/lib/jenkins/jobs/gh-pmix-pr/workspace/out/out
02:14:40 ERROR [cli_stages.c:256:errhandler]:  PMIX server event handler with status = -25
02:14:40 [jenkins03:05363] PMIX ERROR: UNPACK-FAILURE in file /var/lib/jenkins/jobs/gh-pmix-pr/workspace/src/server/pmix_server_ops.c at line 197
02:14:42 [jenkins03:05363] PMIX ERROR: UNPACK-FAILURE in file /var/lib/jenkins/jobs/gh-pmix-pr/workspace/src/server/pmix_server_ops.c at line 197
02:14:42 [jenkins03:05363] PMIX ERROR: UNPACK-FAILURE in file /var/lib/jenkins/jobs/gh-pmix-pr/workspace/src/server/pmix_server_ops.c at line 197
02:14:50 ERROR [pmix_test.c:170:main]: Test exited by a timeout!
02:14:50 ERROR [cli_stages.c:148:cli_cleanup]: rank 0 with state 1 unexpectedly terminated.
02:14:50 ERROR [pmix_test.c:176:main]: Test was aborted!
```

```
02:14:30 ++ timeout -s SIGKILL 1m ./pmix_test -n 4 --ns-dist 3:1 --fence '[b | 0:]' -o /var/lib/jenkins/jobs/gh-pmix-pr/workspace/out/out
02:14:30 ERROR [cli_stages.c:256:errhandler]:  PMIX server event handler with status = -25
02:14:30 [jenkins03:05277] PMIX ERROR: UNPACK-FAILURE in file /var/lib/jenkins/jobs/gh-pmix-pr/workspace/src/server/pmix_server_ops.c at line 197
02:14:30 [jenkins03:05277] PMIX ERROR: UNPACK-FAILURE in file /var/lib/jenkins/jobs/gh-pmix-pr/workspace/src/server/pmix_server_ops.c at line 197
02:14:30 [jenkins03:05277] PMIX ERROR: UNPACK-FAILURE in file /var/lib/jenkins/jobs/gh-pmix-pr/workspace/src/server/pmix_server_ops.c at line 197
02:14:40 ERROR [pmix_test.c:170:main]: Test exited by a timeout!
02:14:40 ERROR [cli_stages.c:148:cli_cleanup]: rank 0 with state 1 unexpectedly terminated.
02:14:40 ERROR [pmix_test.c:176:main]: Test was aborted!
```

```
02:14:20 + eval timeout -s SIGKILL 1m ./pmix_test -n 4 --ns-dist 3:1 --fence '"[0:]"' -o '$OUTDIR/out'
02:14:20 ++ timeout -s SIGKILL 1m ./pmix_test -n 4 --ns-dist 3:1 --fence '[0:]' -o /var/lib/jenkins/jobs/gh-pmix-pr/workspace/out/out
02:14:20 [jenkins03:05185] PMIX ERROR: UNPACK-FAILURE in file /var/lib/jenkins/jobs/gh-pmix-pr/workspace/src/server/pmix_server_ops.c at line 197
02:14:20 [jenkins03:05185] PMIX ERROR: UNPACK-FAILURE in file /var/lib/jenkins/jobs/gh-pmix-pr/workspace/src/server/pmix_server_ops.c at line 197
02:14:20 [jenkins03:05185] PMIX ERROR: UNPACK-FAILURE in file /var/lib/jenkins/jobs/gh-pmix-pr/workspace/src/server/pmix_server_ops.c at line 197
02:14:20 ERROR [cli_stages.c:256:errhandler]:  PMIX server event handler with status = -25
02:14:30 ERROR [pmix_test.c:170:main]: Test exited by a timeout!
02:14:30 ERROR [cli_stages.c:148:cli_cleanup]: rank 0 with state 1 unexpectedly terminated.
02:14:30 ERROR [pmix_test.c:176:main]: Test was aborted!
```

```
02:14:10 ++ timeout -s SIGKILL 1m ./pmix_test -n 4 --ns-dist 3:1 --fence '[db | 0:;1:]' -o /var/lib/jenkins/jobs/gh-pmix-pr/workspace/out/out
02:14:10 [jenkins03:05094] PMIX ERROR: UNPACK-FAILURE in file /var/lib/jenkins/jobs/gh-pmix-pr/workspace/src/server/pmix_server_ops.c at line 197
02:14:10 [jenkins03:05094] PMIX ERROR: UNPACK-FAILURE in file /var/lib/jenkins/jobs/gh-pmix-pr/workspace/src/server/pmix_server_ops.c at line 197
02:14:10 [jenkins03:05094] PMIX ERROR: UNPACK-FAILURE in file /var/lib/jenkins/jobs/gh-pmix-pr/workspace/src/server/pmix_server_ops.c at line 197
02:14:10 [jenkins03:05094] PMIX ERROR: UNPACK-FAILURE in file /var/lib/jenkins/jobs/gh-pmix-pr/workspace/src/server/pmix_server_ops.c at line 197
02:14:20 ERROR [pmix_test.c:170:main]: Test exited by a timeout!
02:14:20 ERROR [cli_stages.c:148:cli_cleanup]: rank 0 with state 1 unexpectedly terminated.
02:14:20 ERROR [pmix_test.c:176:main]: Test was aborted!
```

```
02:14:00 ++ timeout -s SIGKILL 1m ./pmix_test -n 4 --ns-dist 3:1 --fence '[db | 0:;1:3]' -o /var/lib/jenkins/jobs/gh-pmix-pr/workspace/out/out
02:14:00 [jenkins03:05009] PMIX ERROR: UNPACK-FAILURE in file /var/lib/jenkins/jobs/gh-pmix-pr/workspace/src/server/pmix_server_ops.c at line 197
02:14:00 [jenkins03:05009] PMIX ERROR: UNPACK-FAILURE in file /var/lib/jenkins/jobs/gh-pmix-pr/workspace/src/server/pmix_server_ops.c at line 197
02:14:00 [jenkins03:05009] PMIX ERROR: UNPACK-FAILURE in file /var/lib/jenkins/jobs/gh-pmix-pr/workspace/src/server/pmix_server_ops.c at line 197
02:14:00 [jenkins03:05009] PMIX ERROR: UNPACK-FAILURE in file /var/lib/jenkins/jobs/gh-pmix-pr/workspace/src/server/pmix_server_ops.c at line 197
02:14:10 ERROR [pmix_test.c:170:main]: Test exited by a timeout!
02:14:10 ERROR [cli_stages.c:148:cli_cleanup]: rank 0 with state 1 unexpectedly terminated.
02:14:10 ERROR [pmix_test.c:176:main]: Test was aborted!
```

```
02:13:50 ++ timeout -s SIGKILL 1m ./pmix_test -n 4 --ns-dist 3:1 --fence '[db | 0:0-2;1:3]' -o /var/lib/jenkins/jobs/gh-pmix-pr/workspace/out/out
02:13:50 [jenkins03:04899] PMIX ERROR: UNPACK-FAILURE in file /var/lib/jenkins/jobs/gh-pmix-pr/workspace/src/server/pmix_server_ops.c at line 197
02:13:50 [jenkins03:04899] PMIX ERROR: UNPACK-FAILURE in file /var/lib/jenkins/jobs/gh-pmix-pr/workspace/src/server/pmix_server_ops.c at line 197
02:13:50 [jenkins03:04899] PMIX ERROR: UNPACK-FAILURE in file /var/lib/jenkins/jobs/gh-pmix-pr/workspace/src/server/pmix_server_ops.c at line 197
02:13:50 [jenkins03:04899] PMIX ERROR: UNPACK-FAILURE in file /var/lib/jenkins/jobs/gh-pmix-pr/workspace/src/server/pmix_server_ops.c at line 197
02:14:00 ERROR [pmix_test.c:170:main]: Test exited by a timeout!
02:14:00 ERROR [cli_stages.c:148:cli_cleanup]: rank 0 with state 1 unexpectedly terminated.
02:14:00 ERROR [pmix_test.c:176:main]: Test was aborted!
```

**Comment by rhc54** [comment]
I'm beginning to suspect that the tests running in the Jenkins server are stale and may need updating. I found one error and fixed it (the test server was both returning non-success and calling the callback function, which is illegal - if the initial function call fails, then it should return non-success and not call the callback function since the function was never executed). I suspect there are others that may now be getting tripped as the library itself is being more picky.

**Comment by karasevb** [comment]
@rhc54 Please roll back my commit 8f260fd, it does not work and was added mistakenly. I will create new PR against that when all changes are ready

**Comment by rhc54** [comment]
@karasevb No problem - done!

**Comment by rhc54** [comment]
@artpol84 @karasevb It is clear that something is busted on Mellanox Jenkins here. It's been failing everything for a few days now, and it just hangs when we ask for the console output. Can someone please take a look?

**Comment by artpol84** [comment]
@rhc54 I was able to open console, though it was slow (we will check that).

**Comment by karasevb** [comment]
@rhc54 @artpol84 the cause of hangs is the huge size of console output file >2Gb
Log is filled the following output:
```
01:53:16 ERROR [cli_stages.c:256:errhandler]:  PMIX server event handler with status = -25
01:53:16 ERROR [cli_stages.c:256:errhandler]:  PMIX server event handler with status = -25
01:53:16 ERROR [cli_stages.c:256:errhandler]:  PMIX server event handler with status = -25
```

**Comment by rhc54** [comment]
You'll always get that "error" - it just means that the client process terminated. You might want to update that test to ignore it rather than printing it out.

**Comment by rhc54** [comment]
I see why you are getting so many of these outputs - I'll add some protection to eliminate them

**Comment by karasevb** [comment]
@rhc54 Please look, this doesn't allow Jenkins to be happy.
I see the Segmentation fault in `pmix_test` which appears here: [pmix_server_get.c:525](https://github.com/pmix/pmix/pull/354/files#diff-22fc1c8f17317100083a67500cff1ee1R525)

Backtrace:
`cmd: ./pmix_test -n 4 --ns-dist 3:1 --fence "[db | 0:;1:]"`
```
(gdb) bt
#0  0x00007fd317aadad5 in _satisfy_request (nptr=0x7fd310000a40, rank=2, cd=0x0, cbfunc=0x7fd317a86b86 <get_cbfunc>, cbdata=0x7fd310021370, local=0x0)
    at server/pmix_server_get.c:525
#1  0x00007fd317aaf05a in pmix_pending_resolve (nptr=0x7fd310000a40, rank=2, status=0, lcd=0x0) at server/pmix_server_get.c:708
#2  0x00007fd317a9040e in pmix_server_commit (peer=0x7fd31001f410, buf=0x7fd315f81da0) at server/pmix_server_ops.c:261
....
(gdb) l
524	    if (0 < nptr->nlocalprocs) {
525	        *local = true;                                                 <- Segfault here
526	        if (PMIX_RANK_WILDCARD == rank) {
527	            /* use undefined scope */
528	            scope = PMIX_SCOPE_UNDEF;
(gdb) frame 1
#1  0x00007fd317aaf05a in pmix_pending_resolve (nptr=0x7fd310000a40, rank=2, status=0, lcd=0x0) at server/pmix_server_get.c:708
708	            rc = _satisfy_request(nptr, rank, NULL, req->cbfunc, req->cbdata, NULL);
                                                                                      ^^^^
```

**Comment by karasevb** [comment]
@rhc54 There are four PRs which fixes Jenkins failures: [#9](https://github.com/rhc54/pmix-master/pull/9), [#10](https://github.com/rhc54/pmix-master/pull/10),  [#11](https://github.com/rhc54/pmix-master/pull/11),  [#12](https://github.com/rhc54/pmix-master/pull/12)

My Jenkins report below. All of tests uses the GDS hash module.
```
Checking without dstor:
1..14
ok 1 blocking fence w/ data all (./pmix_test -n 4 --ns-dist 3:1 --fence "[db | 0:0-2;1:3]" -o $OUTDIR/out)
ok 2 blocking fence w/ data all (./pmix_test -n 4 --ns-dist 3:1 --fence "[db | 0:;1:3]" -o $OUTDIR/out)
ok 3 blocking fence w/ data all (./pmix_test -n 4 --ns-dist 3:1 --fence "[db | 0:;1:]" -o $OUTDIR/out)
ok 4 non-blocking fence w/o data (./pmix_test -n 4 --ns-dist 3:1 --fence "[0:]" -o $OUTDIR/out)
ok 5 blocking fence w/ data (./pmix_test -n 4 --ns-dist 3:1 --fence "[b | 0:]" -o $OUTDIR/out)
ok 6 non-blocking fence w/ data (./pmix_test -n 4 --ns-dist 3:1 --fence "[d | 0:]" --noise "[0:0,1]" -o $OUTDIR/out)
ok 7 blocking fence w/ data on the same nspace (./pmix_test -n 4 --job-fence -c -o $OUTDIR/out)
ok 8 blocking fence w/o data on the same nspace (./pmix_test -n 4 --job-fence -o $OUTDIR/out)
ok 9 publish (./pmix_test -n 2 --test-publish -o $OUTDIR/out)
ok 10 spawn (./pmix_test -n 2 --test-spawn -o $OUTDIR/out)
ok 11 connect (./pmix_test -n 2 --test-connect -o $OUTDIR/out)
ok 12 resolve peers (./pmix_test -n 5 --test-resolve-peers --ns-dist "1:2:2" -o $OUTDIR/out)
ok 13 key replacement (./pmix_test -n 5 --test-replace 100:0,1,10,50,99 -o $OUTDIR/out)
ok 14 local store (./pmix_test -n 5 --test-internal 10 -o $OUTDIR/out)
+ exit 0
```

File: src/client/pmix_client.c [hunk 1]
```diff
@@ -964,8 +961,14 @@ static void _commitfn(int sd, short args, void *cbdata)
      * that we contributed whatever we had */
     PMIX_PTL_SEND_RECV(rc, &pmix_client_globals.myserver, msgout,
                        wait_cbfunc, (void*)&cb->active);
+    if (PMIX_SUCCESS == rc) {
```
Comment (artpol84) [inline src/client/pmix_client.c]: @rhc54
I'd appreciate if when possible you'd reopen Boris PR, merge it and do the refactoring.

Comment (rhc54) [inline src/client/pmix_client.c]: ??? I don't understand - what does that accomplish?

Comment (artpol84) [inline src/client/pmix_client.c]: Preserves original authorship.

**Comment by rhc54** [comment]
😄  Let's worry about function and just get this thing working

**Comment by artpol84** [comment]
As you can see, we do care about functionality - we provide fixes. And I don't see any problems with us being authors of those fixes no matter how trivial they are.

**Comment by rhc54** [comment]
This is likely ready to go, but will be held pending creation of the v2.0 branch

**Comment by karasevb** [comment]
@rhc54  Jenkins is still unhappy. There is building problem, which appeared after commit edb8ec5f842588b15416698dea806127fd97d839. This is due to the added a file `common/pmix_data.c`.

**Comment by rhc54** [comment]
Yes, I'll need to merge and fix the conflicts. I'll get to it after v2.0 is released since we can't commit before then anyway.

**Comment by abouteiller** [comment]
Is this also supposed to help interoperate with pmi2 and/or pmi ? I have a Slurm install that produces PMI2 info I can try if that.

**Comment by rhc54** [comment]
@abouteiller The backward compatibility to PMI-1 and PMI-2 is already in the library. This PR deals with compatibility between PMIx versions.

**Comment by rhc54** [comment]
@artpol84 @karasevb @jjhursey I have this updated and ready to merge to master. However, I'd appreciate some testing in advance to ensure I got it right, even though it passes Jenkins.

**Comment by artpol84** [comment]
@karasevb please run through our MTT

**Comment by artpol84** [comment]
@rhc54 I don't understand the list of commits.
Where is @karasevb commits? Where is the history?

**Comment by artpol84** [comment]
The following PR was merged here, but I don't see any of those commits:
https://github.com/rhc54/pmix-master/pull/7/commits

**Comment by artpol84** [comment]
@karasevb is working on https://github.com/open-mpi/ompi/pull/3787 and found that his changes (https://github.com/karasevb/master/commit/6c8875e3ef4c70e798ed1948cac4fe065c71c538) was part of his PR against this branch they are not in this PR anymore.
This lead me to check this PR again and I'm totally confused. @rhc54 have you rejected our changes?

**Comment by artpol84** [comment]
I checked d95dc6275d6b94b038f7855c417de2f1f38aa568 and I see @karasevb changes there. This is not right, @rhc54. We had this discussion before - **please do not change the authorship of our work**!

@jladd-mlnx

**Comment by rhc54** [comment]
I did _not_ change the authorship, so calm down. It is in all the commit messages. However, we have to rebase the PR in order to make it manageable. Otherwise, every single commit to master generates a conflict with every commit in the sequence.

If you want to fight thru that mess every time, that's your problem - not mine

**Comment by rhc54** [comment]
@jjhursey and I discussed the situation with respect to this PR, master, and the v2.0 branch. The current status of v2.0.1 is looking quite solid, and so we don't anticipate major fixes required there. Thus, we can reasonably work thru a change to master that causes a major divergence between the branches.

Even though this PR isn't fully functional, it does pass CI (since that is single-node) and it does work multi-node inside OMPI if the "hash" GDS component is used. So perhaps the best path forward is for this PR to be committed as-is, and then I'll file a separate issue for the multi-node dstore problem so that @karasevb can resolve it and directly commit the fix as his schedule permits. Meantime, I'll send a note to the mailing list warning of the limitation in master.

**Comment by artpol84** [comment]
@rhc54, commit message you referred is not equivalent to normal commits for the following reasons:
- it doesn't allow to judge about the changes;
- does not represented in @karasevb repo statistics (https://github.com/pmix/pmix/graphs/contributors). This was a visible contribution - not a couple of lines;

If you have problems with merging I expect you to talk to us first before deciding what to do with our contribution.

**Comment by jjhursey** [comment]
We should move this to the mailing list for discussion so we are all on the same page moving forward with the project. This PR has merged so nothing we can do about it now. Hopefully we can agree on a set of guiding principles to make this process a little less bumpy in the future.

**Comment by artpol84** [comment]
@jjhursey, thanks! :+1: for the discussion.

**Pull request merged**